cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java

```diff
@@ -217,13 +223,21 @@ private static boolean subnetExists(List<String> subnets, String subnet) {
 
   // subnets are identified as
   // "https://www.googleapis.com/compute/v1/projects/<project>/regions/<region>/subnetworks/<name>"
-  // a subnet in the same region as the dataproc cluster must be chosen
-  private static String chooseSubnet(String network, List<String> subnets, String zone) {
+  // a subnet in the same region as the dataproc cluster must be chosen. If a subnet name is provided then the subnet
+  // will be choosen and the region will be picked on basis of the given zone. If a subnet name is not provided then
+  // any subnetwork in the region of the given zone will be picked.
+  private static String chooseSubnet(String network, List<String> subnets, @Nullable String subnet, String zone) {
     // zones are always <region>-<letter>
     String region = zone.substring(0, zone.lastIndexOf('-'));
-    for (String subnet : subnets) {
-      if (subnet.contains(region + "/subnetworks")) {
-        return subnet;
+    for (String currentSubnet : subnets) {
+      // if a subnet name is given then get the region of that subnet based on the zone
+      if (subnet != null) {
+        if (!currentSubnet.endsWith("subnetworks/" + subnet)) {
```

Comment:
this can be combined with the previous line.

Reply:
Done .